gremlin-server/src/main/java/org/apache/tinkerpop/gremlin/server/handler/MultiTaskSession.java

```diff
@@ -0,0 +1,283 @@
+    private ScheduledFuture<?> requestCancelFuture;
+    private Bindings bindings;
+
+    /**
+     * Creates a new {@code MultiTaskSession} object providing the initial starting {@link SessionTask} that gets
+     * executed by the session when it starts.
+     *
+     * @param initialSessionTask The tasks that starts the session.
+     * @param sessionId The id of the session
+     * @param sessions The session id to {@link Session} instances mapping
+     */
+    public MultiTaskSession(final SessionTask initialSessionTask, final String sessionId,
+                     final ConcurrentMap<String, Session> sessions) {
+        super(initialSessionTask, sessionId, false, sessions);
+
+        queue = new LinkedBlockingQueue<>(initialSessionTask.getSettings().maxSessionTaskQueueSize);
+
+        // using a global function cache is cheaper than creating a new on per session especially if you have to
+        // create a lot of sessions. it will generate a ton of throw-away objects. mostly keeping the option open
+        // to not use it to preserve the ability to use the old functionality if wanted or if there is some specific
+        // use case with sessions that needs it. if we wanted this could eventually become a per-request option
+        // so that the client could control it as necessary and get scriptengine isolation if they need it.
+        if (initialSessionTask.getSettings().useCommonEngineForSessions)
+            scriptEngineManager = initialSessionTask.getGremlinExecutor().getScriptEngineManager();
+        else
+            scriptEngineManager = initializeGremlinExecutor(initialSessionTask).getScriptEngineManager();
+
+        scheduledExecutorService = initialSessionTask.getScheduledExecutorService();
+        submitTask(initialSessionTask);
+    }
+
+    /**
+     * Gets the script engine specific to this session which is dependent on the
+     * {@link Settings#useCommonEngineForSessions} configuration.
+     */
+    @Override
+    public GremlinScriptEngine getScriptEngine(final SessionTask sessionTask, final String language) {
+        return scriptEngineManager.getEngineByName(language);
+    }
+
+    @Override
+    public boolean isAcceptingTasks() {
+        return !ending.get();
+    }
+
+    @Override
+    public boolean submitTask(final SessionTask sessionTask) throws RejectedExecutionException {
+        try {
+            return isAcceptingTasks() && queue.add(sessionTask);
+        } catch (IllegalStateException ise) {
+            final String msg = String.format("Task %s rejected from session %s",
+                    sessionTask.getRequestMessage().getRequestId(), getSessionId());
+            throw new RejectedExecutionException(msg);
+        }
+    }
+
+    @Override
+    public void run() {
+        // allow the Session to know about the thread that is running it - the thread really only has relevance
+        // once the session has started.
+        this.sessionThread = Thread.currentThread();
```

Comment:
How do we ensure that all requests in the same session are run using the same thread (since trx might be thread bound)?

Reply:
Adding a comment as follows to further clarify:

```java
        // allow the Session to know about the thread that is running it which will be a thread from the gremlinPool
        // the thread really only has relevance once the session has started. this thread is then held below in the
        // while() loop awaiting items to arrive on the request queue. it will hold until the session is properly
        // closed by the client or through interruption/error
```

Reply:
ah, in that case, if the default worker pool size is 8, then, we would not be able to process 9th new concurrent session? Because all the threads will be stuck serving the existing sessions? This will also impact single-session aka session-less requests.

I understand that this is not different from current behaviour but today a single session request does not get impacted by multi session query because they share different worker thread pools. We are changing that assumption here. A customer who is relying on the behaviour above to run longer/multiple queries on sessions and low latency faster query on session-less will see a regression from this change. We should document that somewhere.

In future, we should fix this limitation. This is the "starvation by session requests" I was referring to in my email reply over the dev mailing list. There are multiple ways to fix this but perhaps we should treat it as out of scope for this PR.

Reply:
so, the basic problem is that the TinkerPop API doesn't really let us move a transaction context across threads. There is a notion of the "threaded transactions" which does allow this but no one really implements that besides JanusGraph. will think about this some more but i think that ultimately it will expand scope beyond what this pull request should get into. there is already some documentation on the potential impact to performance that you describe but i will review again.